### ISSUE TITLE
Can stylelint work with Material UI's theme?

### ISSUE THREAD

**Issue by imakou** [stylelint/stylelint#4002]
Hi all,

I'd like to use StyleLint to lint CSS-in-JS(object literals) with Material UI.
I use StyleLint with VS Code. It works well with *.css files when I save the files, but it doesn't work with my JSX file.

In my .stylelintrc
```
{
  "extends": "stylelint-config-recommended",
  "rules": {
    "color-no-invalid-hex": true,
    "color-hex-case": "upper",
    "color-hex-length": "long",
  }
}
```
It seems that the Stylelint doesn't work on the file.
![image](https://user-images.githubusercontent.com/1507950/55284280-6987d500-5341-11e9-8a26-7151a93be8a0.png)

How to make Stylelint to watch my jsx files as well?

Thanks.

**Comment by jeddy3** [comment]
> How to make Stylelint to watch my jsx files as well?

stylelint can extract embedded CSS from within CSS-in-JS object & template literals in JSX files. For example:

```js
export default (
  <button
    style={{ color: "#fff" }}
  ></button>
);`
```

```js
import styled from 'styled-components';
export default styled.div`
  padding-left: 10px;
  padding: 20px;
`;
```

The `theme` variable in your code feels more like a configuration object and not really CSS-in-JS so it's doubtful stylelint will lint that.

**Comment by litera** [comment]
@jeddy3 I would ask you to reopen this issue. The example original poster provided doesn't really exemplify the actual CSS-in-JS styling possibilities of the MUI, so he falsely let you believe that MUI theming is just some sort of configuration object.

This is more in line what's possible and needs to be parsed and linted.

## `sx` component property ([MUI docs link](https://mui.com/system/getting-started/the-sx-prop/))
This is a kind of a DOM element inline styles definition
```
// number only for dimensions, which get's translated to pixels where 1=8px (default that can be changed)
<Box sx={{ margin: 1 }} /> 

// MUI supports shorthand CSS prop names (not mandatory to support in stylelint)
// https://mui.com/system/getting-started/the-sx-prop/#spacing
<Box sx={{ m: 1 }} />

// Theme configuration aware properties
<Box sx={{ margin: 1, gap: 1 }} /> // both equivalent to '8px'
<Box sx={{ borderRadius: 1 }} /> // equivalent to '4px'
<Box sx={{ width: 0.5 }} /> // equivalent to '50%'
<Box sx={{ width: 50 }} /> // equivalent to '50px'
<Box sx={{ typography: 'body1' }} /> // theme defines various named typography variants with full styling definitions
<Box sx={{ boxShadow: 1 }} /> // equivalent to theme => theme.shadows[1]
<Box sx={{ color: 'primary.main' }} /> // equivalent to theme => theme.palette.primary.main

// MUI theme defines responsive breakpoints and one can provide different CSS prop values for various screen sizes
<Box sx={{ m: { xs: 1, sm: 1.5, md: 2, lg: 2.5, xl: 3 } }} />

// Style definition as a callback
<Box sx={{ m: ({ theme, ownerState }) => theme.spacing(1) }} />
// or
<Box sx={({ theme, ownerState}) => ({ m: theme.spacing(1), p: theme.spacing(1, 2) })} />

// Array values
<Box sx={[{ m: 1 }, hasError && { color: '#f00' }]} />

// Nested selectors
<Box sx={{ m: 1, ':hover': { color: '#00f' } }} />
// Selectors being object prop names can of course (as per TS/JS) be interpolated
<Box sx={{ m: 1, [`&.${stateClasses.disabled}`]: { opacity: 0.5, pointerEvents: 'none' } }} />
```

## `styled` component generation ([MUI docs link](https://mui.com/system/styled/))
MUI uses emotion styled utility but extends it with the upper changes that are available on `sx`. All of those. It's usualy also called as a function:
```
const ComponentRoot = styled(Box, [optionsObj])({
  // component root styles
  '::before': {
    // nested styles
  },
  [`:has(${itemClasses.selected})`]: {
    // also complex interpolated nested selectors
    ':hover': {
      // even more nesting
    }
  }
});

// also callback styles
const ComponentRoot = styled(Box)(({ theme, ownerState }) => ({
  margin: theme.spacing(1,2,3,4),
  transition: theme.transitions.create('color'),
  ':hover': {
     color: theme.palette.secondary.main
  }
}));
```

## Component theming ([MUI docs link]())
MUI material components can be re-themed by providing style overrides or developers can wrote their own components that can be themed when used by other teams.
```
const themeOverides = {
  MuiButton: {
    styleOverrides: {
      root: {
        // button root "slot" style overrides as a simple object
      },
      secondary: ({theme, owserState}) => ({
        // button styles when variant "secondary" are provided as a callback
      })
    }
  }
}
```
All these stales can be very complex providing nested selectors, use theming variables/functions etc. When the theme is created these themes can be provided to it and the default MUI components will be restyled as per these style overrides.

## The most important
All these styles are provided as CSS-in-JS/TS, and especially when using Typescript there are very well defined types for all of these so you get code hinting, TS linting for invalid CSS prop values etc. But stylelint could support this syntax so it would be able to lint these styles according to preset rules. I'm not even going to think about browser compatibility issue linting (as per eslint-plugin-compat) that uses *browserslist* infromation coupled with *caniuse* data.
